contracts/src/LightClientV2.sol

```diff
@@ -0,0 +1,254 @@
+// SPDX-License-Identifier: UNLICENSED
+
+pragma solidity ^0.8.0;
+pragma experimental ABIEncoderV2;
+
+import { OwnableUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import { Initializable } from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import { UUPSUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+import { BN254 } from "bn254/BN254.sol";
+import { IPlonkVerifier } from "./interfaces/IPlonkVerifier.sol";
+import { PlonkVerifier } from "./libraries/PlonkVerifier.sol";
+import { LightClientStateUpdateVK as VkLib } from "./libraries/LightClientStateUpdateVK.sol";
+
+/// @notice A light client for HotShot consensus. Keeping track of its finalized states in safe,
+/// authenticated ways.
+contract LightClientV2 is Initializable, OwnableUpgradeable, UUPSUpgradeable {
+    // === Events ===
+    //
+    // @notice Notify a new epoch is starting
+    event EpochChanged(uint64);
+
+    /// @notice upgrade event when the proxy updates the implementation it's pointing to
+    event Upgrade(address implementation);
+
+    // === Constants ===
+    //
+    /// @notice System parameter: number of blocks per epoch
+    /// @dev This variable cannot be made immutable due to how UUPS contracts work. See
+    /// https://forum.openzeppelin.com/t/upgradable-contracts-instantiating-an-immutable-value/28763/2#why-cant-i-use-immutable-variables-1
+    uint32 public blocksPerEpoch;
+    /// @notice A simple way to track contract versions
+    uint32 public immutable VERSION = 2;
+
+    // === Storage ===
+    //
+    /// @notice genesis block commitment
+    LightClientState public genesisState;
+    /// @notice global storage of the finalized HotShot's light client state
+    LightClientState public finalizedState;
+    /// @notice current (finalized) epoch number
+    uint64 public currentEpoch;
+    /// @notice The commitment of the stake table used in current voting (i.e. snapshot at the start
+    /// of last epoch)
+    bytes32 public votingStakeTableCommitment;
+    /// @notice The quorum threshold for the stake table used in current voting
+    uint256 public votingThreshold;
+    /// @notice The commitment of the stake table frozen for change (i.e. snapshot at the start of
+    /// last epoch)
+    bytes32 public frozenStakeTableCommitment;
+    /// @notice The quorum threshold for the frozen stake table
+    uint256 public frozenThreshold;
+
+    // TODO adding this field to the LightClientState struct does not seem to work. Why? See
```

Comment:
@philippecamacho I think this makes sense because the members of the struct variable are tightly packed in storage so adding a new field to a struct it's similar to adding a new state variable, and is only safe at the end. because that memory is already used by the variable that follows the variable of type LightClientState.

Reply:
So in order to be able to add new field to the struct I guess we should add empty slots to this struct to be used in the future using the [gap](https://github.com/EspressoSystems/espresso-sequencer/pull/1079#discussion_r1489886269) approach.

Or we may put the struct into a map as contract updates work well in this [case](https://github.com/EspressoSystems/espresso-sequencer/blob/fb11109a5c14a48eba4aa2dd115865a55f43ad79/contracts/demo/upgradeDemo/DemoBoxV2.sol#L30).

Reply:
Comment updated in f8c63d42830c9f24a11fd81a1ccf8d38a2db15af.

Reply:
Finally, aftet discussing with @alysiahuggins we decided to store the light client state struct into a map. See https://github.com/EspressoSystems/espresso-sequencer/pull/1079#discussion_r1494738481